sycl/plugins/esimd_emulator/pi_esimd_emulator.cpp

```diff
@@ -1002,45 +1017,56 @@ pi_result piMemBufferCreate(pi_context Context, pi_mem_flags Flags, size_t Size,
     return PI_INVALID_VALUE;
   }
 
-  cm_support::CmBuffer *CmBuf = nullptr;
-  cm_support::SurfaceIndex *CmIndex;
-
-  int Status = Context->Device->CmDevicePtr->CreateBuffer(
-      static_cast<unsigned int>(Size), CmBuf);
-
-  if (Status != cm_support::CM_SUCCESS) {
-    return PI_OUT_OF_HOST_MEMORY;
+  // Flag & HostPtr argument sanity check
+  if (!Context->checkSurfaceArgument(Flags, HostPtr)) {
+    return PI_INVALID_OPERATION;
   }
 
-  Status = CmBuf->GetIndex(CmIndex);
-  const std::lock_guard<sycl::detail::SpinLock> Lock{*PiESimdSurfaceMapLock};
-  assert(PiESimdSurfaceMap->find((unsigned int)CmIndex->get_data()) ==
-             PiESimdSurfaceMap->end() &&
-         "Failure from CM-managed buffer creation");
-
-  // Initialize the buffer with user data provided with 'HostPtr'
-  if ((Flags & PI_MEM_FLAGS_HOST_PTR_USE) != 0) {
-    if (HostPtr != nullptr) {
-      Status =
-          CmBuf->WriteSurface(reinterpret_cast<const unsigned char *>(HostPtr),
-                              nullptr, static_cast<unsigned int>(Size));
+  char *MapBasePtr = nullptr;
+  cm_buffer_ptr CmBuf;
+  cm_support::SurfaceIndex *CmIndex;
+  std::lock_guard<std::mutex> Lock(Context->CmSurfaceManageLock);
+  int Status = cm_support::CM_FAILURE;
+  bool IsHostMem = false;
+
+  if (Flags & PI_MEM_FLAGS_HOST_PTR_USE) {
+    Status = Context->Device->CmDevicePtr->CreateBufferUP(
+        static_cast<unsigned int>(Size), HostPtr, CmBuf.HostPtr);
+    CmBuf.HostPtr->GetIndex(CmIndex);
+    IsHostMem = true;
+  } else {
+    Status = Context->Device->CmDevicePtr->CreateBuffer(
+        static_cast<unsigned int>(Size), CmBuf.GPUPtr);
+    CmBuf.GPUPtr->GetIndex(CmIndex);
+
+    if (Flags & PI_MEM_FLAGS_HOST_PTR_COPY) {
+      CmBuf.GPUPtr->WriteSurface(
+          reinterpret_cast<const unsigned char *>(HostPtr), nullptr,
+          static_cast<unsigned int>(Size));
     }
   }
 
-  auto HostPtrOrNull =
-      (Flags & PI_MEM_FLAGS_HOST_PTR_COPY) ? nullptr : pi_cast<char *>(HostPtr);
+  if (Status != cm_support::CM_SUCCESS) {
+    return PI_INVALID_OPERATION;
+  }
+
+  MapBasePtr =
+      pi_cast<char *>(cm_support::get_surface_base_addr(CmIndex->get_data()));
 
   try {
-    *RetMem =
-        new _pi_buffer(Context, HostPtrOrNull, CmBuf,
-                       /* integer buffer index */ CmIndex->get_data(), Size);
+    *RetMem = new _pi_buffer(Context, MapBasePtr, CmBuf, CmIndex->get_data(),
```

Comment:
I'm not going to hold this PR, but it seems there is some redundancy - CmBuf has HostPtr and GPUPtr fields, yet IsHostMem is also passed to the constructor. Let's discuss this on a sync meeting.